### PR TITLE
PLANET-2479: Gallery block

### PR DESCRIPTION
**Pull request by tamlyn**
## Task

> We combine and retire the 3 column and carousel blocks, and add functionality for a grid of images.
https://jira.greenpeace.org/browse/PLANET-2479

## Admin

The description field is currently only used by the "3 columns" layout. Should it be hidden for the other layouts, or should the templates be updated to support an optional description?

## Frontend

The existing layouts are unchanged. The new grid layout shows images at a fixed aspect ration of 3:2 in rows of 2, 3 or 4 images depending on the window width. This means the last row will be incomplete unless the total number of images in a multiple of 12.

### Desktop
www planet4 test_gallery_

### Tablet
www planet4 test_gallery_ 1

### Mobile
www planet4 test_gallery_ iphone 5_se

The mockup on the ticket shows a gap of 30px between each image in the grid but this looks strange on a mobile where the gap between the content and the edge of the screen is only 15px. I've changed the gap to 15px everywhere but I could make it 30px on everything but mobile.

### CODE REVIEW

File: classes/controller/blocks/class-gallery-controller.php [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+				self::LAYOUT_GRID          => 'medium',
+			];
+
+			foreach ( $exploded_images as $image_id ) {
+				$image_size = $image_sizes[ $fields['gallery_block_style'] ];
+				$image_size = $image_sizes[ $fields['gallery_block_style'] ];
+				$image_data = [];
+
+				$image_data_array           = wp_get_attachment_image_src( $image_id, $image_size );
+				$image_data['image_src']    = $image_data_array[0];
```
Comment (angtheod) [inline classes/controller/blocks/class-gallery-controller.php]: Add a check in case `wp_get_attachment_image_src` returns false.

**Comment by angtheod** [comment]
Overall looks good. Left couple comments. Also, needs to resolve Css conflicts.

File: classes/controller/blocks/class-gallery-controller.php [hunk 2]
```diff
@@ -156,8 +156,14 @@ public function prepare_data( $fields, $content = '', $shortcode_tag = 'shortcak
 			$gallery_style       = $fields['gallery_block_style'] ?? static::LAYOUT_SLIDER;
 			$gallery_title       = $fields['gallery_block_title'] ?? '';
 			$gallery_description = $fields['gallery_block_description'] ?? '';
+			$images				 = [];
+
+			if (isset($fields['multiple_image'])) {
```
[the remark this answers is not in the document]
Comment (angtheod) [reply classes/controller/blocks/class-gallery-controller.php]: Alternatively we could place the whole foreach within this if and avoid the else part. But that is fine as well.

File: classes/controller/blocks/class-gallery-controller.php [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+		 * @param \WP_Post $post The attachment record in the database.
+		 *
+		 * @return array Final array of form fields to use.
+		 */
+		public function add_image_attachment_fields_to_edit( $form_fields, $post ) {
+
+			// Add a Credit field.
+			$form_fields['credit_text'] = [
+				// phpcs:ignore WordPress.WP.I18n.MissingArgDomain
+				'label' => __( 'Credit' ),
```
Comment (kirdia) [inline classes/controller/blocks/class-gallery-controller.php]: @pablocubico please add the 'planet4-blocks-backend' domain and remove the phpcs comment.

File: classes/controller/blocks/class-gallery-controller.php [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+				'input' => 'text', // this is default if "input" is omitted.
+				'value' => get_post_meta( $post->ID, '_credit_text', true ),
+				// phpcs:ignore WordPress.WP.I18n.MissingArgDomain
```
Comment (kirdia) [inline classes/controller/blocks/class-gallery-controller.php]: same here for the textdomain and the comment

**Comment by comzeradd** [comment]
This has many unrelated commits. It should be either squashed during merge into one commit or rebased.

**Pull request merged**